### PR TITLE
Imgui async render while paused

### PR DESCRIPTION
**Pull request by Hibyehello**
This allows the user to interact with Imgui while the game has been paused. I am looking into also using the async rendering when fps is slow for when tasing.

### CODE REVIEW

File: Source/Core/VideoCommon/RenderBase.cpp [hunk 1]
```diff
@@ -1481,6 +1485,87 @@ void Renderer::Swap(u32 xfb_addr, u32 fb_width, u32 fb_stride, u32 fb_height, u6
   }
 }
 
+void Renderer::UISwap()
+{
+  if (m_xfb_entry)
```
Comment (phire) [inline Source/Core/VideoCommon/RenderBase.cpp:1490]: No reason to skip rendering the whole UI when you don't have an XFB. 
You should be able to skip just the `RenderXFBToScreen` and imgui will render over a blank screen.

File: Source/Core/VideoCommon/RenderBase.cpp [hunk 2]
```diff
@@ -1481,6 +1485,87 @@ void Renderer::Swap(u32 xfb_addr, u32 fb_width, u32 fb_stride, u32 fb_height, u6
...
+
+    g_shader_cache->RetrieveAsyncShaders();
+    g_vertex_manager->OnEndFrame();
+    BeginImGuiFrame();
+
+    // We invalidate the pipeline object at the start of the frame.
+    // This is for the rare case where only a single pipeline configuration is used,
+    // and hybrid ubershaders have compiled the specialized shader, but without any
+    // state changes the specialized shader will not take over.
+    g_vertex_manager->InvalidatePipelineObject();
```
Comment (phire) [inline Source/Core/VideoCommon/RenderBase.cpp:1543]: Shouldn't do this on UI swap

File: Source/Core/VideoCommon/RenderBase.cpp [hunk 3]
```diff
@@ -1481,6 +1485,87 @@ void Renderer::Swap(u32 xfb_addr, u32 fb_width, u32 fb_stride, u32 fb_height, u6
...
+
+    if (m_force_reload_textures.TestAndClear())
+    {
+      g_texture_cache->ForceReload();
```
Comment (phire) [inline Source/Core/VideoCommon/RenderBase.cpp:1547]: Shouldn't do this on UI swap

File: Source/Core/VideoCommon/RenderBase.cpp [hunk 4]
```diff
@@ -1481,6 +1485,87 @@ void Renderer::Swap(u32 xfb_addr, u32 fb_width, u32 fb_stride, u32 fb_height, u6
...
+    }
+    else
+    {
+      // Flush any outstanding EFB copies to RAM, in case the game is running at an uncapped frame
+      // rate and not waiting for vblank. Otherwise, we'd end up with a huge list of pending
+      // copies.
+      g_texture_cache->FlushEFBCopies();
```
Comment (phire) [inline Source/Core/VideoCommon/RenderBase.cpp:1554]: Shouldn't do this on UI swap

File: Source/Core/VideoCommon/RenderBase.cpp [hunk 5]
```diff
@@ -1481,6 +1485,87 @@ void Renderer::Swap(u32 xfb_addr, u32 fb_width, u32 fb_stride, u32 fb_height, u6
...
+    }
+
+    // Handle any config changes, this gets propagated to the backend.
+    CheckForConfigChanges();
```
Comment (phire) [inline Source/Core/VideoCommon/RenderBase.cpp:1558]: Shouldn't do this on UI swap

File: Source/Core/VideoCommon/RenderBase.cpp [hunk 6]
```diff
@@ -1481,6 +1485,87 @@ void Renderer::Swap(u32 xfb_addr, u32 fb_width, u32 fb_stride, u32 fb_height, u6
   }
 }
 
+void Renderer::UISwap()
```
Comment (phire) [inline Source/Core/VideoCommon/RenderBase.cpp:1488]: `UISwap` contains a bunch of copy/pasted code from `Swap`
This common code should be extracted out into a common function that both will call.

**Comment by iwubcode** [comment]
I don't understand the need for doing the "ContentLocking" or the smart-pointer approach (I've explained why I don't think it's a great idea to implement our own smart pointer).

I would imagine all that needs to happen is to stop the CPU from sending additional commands (which is what pause naturally does) and then send the same state that the software received from the last frame as if the software received it again.  I would also avoid calling anything"UI" because there are multiple pieces of logic that could use this (I've talked about FreeLook before).

**Comment by phire** [comment]
> I don't understand the need for doing the "ContentLocking"

The goal is to extend this to inserting extra frames when the game is running too slowly (or potentially the user is single-stepping though code). This introduces an edge case where a game might invalidate an XFB copy and then a "UISwap" gets inserted before the game gets around to presenting it.

Content locking ensures the UI swaps will always continue to show the previously swaped frame until the game next executes a swap.

The change to shared pointers allows components (such as Renderer) to keep a copy of a texture outside the texture cache, without having to add even more complexity to the invalidation logic. In-fact this change already simplifies the invalidation logic. 

*(I also plan to use this shared-ptr + content locking capability in the future for graphics debugger functionality)*

> (I've just reiterate, I don't think it's a great idea to implement our own smart pointer).

Yes, which is why this version just uses regular std::shared_ptr (see `using RcTcacheEntry = std::shared_ptr<TCacheEntry>;`)

> I would imagine all that needs to happen is to stop the CPU from sending additional commands (which is what pause naturally does)

Yes, but this also has to work without the CPU being paused. 

> You would need to send TextureCache a command to tell it not to age out any textures however.

Which is exactly what the ContentLock is, just on a fine-grained per texture bases.

**Comment by phire** [comment]
FifoCI errors are caused by this assert:

```
VideoCommon/TextureCacheBase.cpp:75 E[Video]: Warning: An error occurred.
  Texture cache destroyed before TCacheEntry was destroyed
```

Fixing this requires making renderer reset it's `m_xfb_entry` before texture cache shuts down. 

Should be as simple as reordering `ShutdownShared` so render is shutdown before texture cache, and then adding `m_xfb_entry.reset()` to `Renderer::Shutdown`

**Comment by iwubcode** [comment]
> This introduces an edge case where a game might invalidate an XFB copy and then a "UISwap" gets inserted before the game gets around to presenting it.

See my edit.  Didn't see you responded, sorry.

I don't know what you mean by "extra frames" but I don't think there's a need for all this complexity.  Just re-render the current frame.  XFB should happen naturally like it does for how Dolphin normally runs.

> Yes, but this also has to work without the CPU being paused.

I don't recall the specifics with how the CPU hands off to say there's another frame.  But if a VI is now getting produced that calls `Swap()`, we'd instead just queue up the call and pull them off in another thread.  My idea remains whether the CPU is paused or not.

We're touching all these internals and making these sweeping changes to have a different function point with shared code, that doesn't seem necessary to me.  Retain the previous data that called `Swap()` and trigger it again, whenever desired.

**Comment by phire** [comment]
> Just re-render the current frame. XFB should happen naturally like it does for how Dolphin normally runs.

Yes, it *almost* works, and it's what the first version of this code did. 

However, the call to `GetXFBTexture` isn't 100% guaranteed to be repeatable. Depending on exactly what the game does between it's real swap and a call this fake UI swap, the state of texture cache might have changed and it `GetXFBTexture` with the exact same parameters might return a different result. 

This might result in a corrupted XFB being viable, or the next frame's XFB frame showing up early (which is not what we want when debugging or doing a TAS).

> We're touching all these internals and making a number of changes to have a different function point with shared code, that doesn't seem necessary to me.

I admit it's a large change to texture cache, which could technically be avoided for this exact PR (with just some weird edge cases). 

But that just means I'll just pull these texture cache changes in with a future PR anyway. Might as well pull them in now.

**Comment by iwubcode** [comment]
Sorry, I think I'm coming off a little strong here.

Let me take a step back.... I think there's a handful of problems and a couple of solutions.  I'm not sure my last comment really captured that.

**Problems**

1.  I want to interact with the UI while the game is paused.
2. I want to interact with FreeLook or other graphical features while the game is paused
3. I want to improve frame pacing  and do other stuff in https://bugs.dolphin-emu.org/issues/12852

**Solutions**

* (A) Async present of the screen.  All you need for this is the XFB.  This solves (1) and (3) but not (2).
* (B) Re-draw the last frame using the previous XF/BPmem data.  This solves (1) and (2) but not (3).

I was pushing for (B) because it fits in with what I'd like to do and solves @Hibyehello 's UI request.  But it _doesn't_ solve (3) which I think is what you all are wanting to resolve.

For (A), can we not just make a copy of the XFB and store it in a queue or something?

**Comment by iwubcode** [comment]
Discussed on Discord with @phire .  We agree both (A) and (B) should be done.  This is (A) effectively.

Sorry for making a racket.  I saw the changes to TCache and in my head thought we could have a simpler solution for async present.  But I wasn't thinking of performance.  This will be better in the long run, even if it adds a bit more complexity with the locking.  Even then, the shared-ptrs _reduced_ some complexity, so it works out.

File: Source/Core/VideoCommon/TextureCacheBase.cpp [hunk 7]
```diff
@@ -1739,30 +1772,29 @@ static void GetDisplayRectForXFBEntry(TextureCacheBase::TCacheEntry* entry, u32
...
   }
 
-  // Do we currently have a version of this XFB copy in VRAM?
-  TCacheEntry* entry = GetXFBFromCache(address, width, height, stride);
+  // Do we currently have a mutable version of this XFB copy in VRAM?
+  RcTcacheEntry entry = GetXFBFromCache(address, width, height, stride);
   if (entry)
   {
-    if (entry->is_xfb_container)
+    if (entry->is_xfb_container && !entry->IsLocked())
```
Comment (iwubcode) [inline Source/Core/VideoCommon/TextureCacheBase.cpp:1791]: This `!entry->IsLocked()` should be moved up to the `if (entry ) ` above

**Comment by Pokechu22** [comment]
I haven't looked into it in detail, but didn't the recent imgui/implot changes allow for imgui to update without the game rendering a new frame? Or am I confused (possibly because I was looking at the hardware fifoplayer, which might have been re-presenting the same XFB copy for many frames in a row/several seconds)?

**Comment by Hibyehello** [comment]
> I haven't looked into it in detail, but didn't the recent imgui/implot changes allow for imgui to update without the game rendering a new frame?

As it currently stands in master Imgui gets frozen as soon as you pause the game.

**Comment by JMC47** [comment]
Is this worth keeping open over pursuing a more proper fix to async presentation?

**Comment by phire** [comment]
No. 

I'll open a PR with my work when it's ready.

**Pull request closed without merging**